### PR TITLE
Pr/component/get children

### PR DESCRIPTION
**Pull request by Onset**
Funkce pro získání potomků a oprava ukázkového kódu.

### CODE REVIEW

File: theme/views/front-page.latte [hunk 1]
```diff
@@ -9,7 +9,13 @@
 		<h3>Example of front-end component. Viz <code>scripts/components/Example.ts</code>:</h3>
 		<button class="example-child">Click me</button>
 	</div>
-	{initComponent Example, [ place => '#example-component-placeholder', data => 'Hello world ' . date(c) ]}
+	{initComponent Example, [
+		place => '#example-component-placeholder',
+		data => [
+			name => 'Krasavec',
```
Comment (ViliamKopecky) [inline theme/views/front-page.latte]: 👌😂

File: theme/scripts/components/Component.ts [hunk 2]
```diff
@@ -26,6 +26,23 @@ export default class Component<D> {
 		this.init()
 	}
 
+	protected getChild<C extends HTMLElement>(
+		selector: string,
+		F: Constructor<C>,
```
Comment (ViliamKopecky) [inline theme/scripts/components/Component.ts]: default value to `HTMLElement` is not wanted?

File: theme/scripts/components/Component.ts [hunk 3]
```diff
@@ -26,6 +26,23 @@ export default class Component<D> {
...
+		parent: HTMLElement = this.el
+	): C {
+		const child = parent.querySelector(selector)
+
+		if (!(child instanceof F)) {
+			throw new ComponentInitializationError()
```
Comment (ViliamKopecky) [inline theme/scripts/components/Component.ts]: Method can be called out of `init`, so the error naming (`...InitializationError`) is not accurate 🤡

File: theme/scripts/components/Component.ts [hunk 4]
```diff
@@ -4,18 +4,18 @@ export interface NamedComponent {
...
-export default class Component<D> {
-	protected readonly el: HTMLElement
+export default class Component<D, E extends HTMLElement = HTMLElement> {
+	protected readonly el: E
 	protected readonly data: D
 
 	getListeners = (): EventListeners => []
 
-	constructor(element: HTMLElement, data: D) {
+	constructor(element: E, data: D) {
```
Comment (enzy) [inline theme/scripts/components/Component.ts]: I'm getting lost in the naming convention for `E`,  `C`, `D`, `F` ... what these stand for and can we name them fully?

File: theme/scripts/components/Component.ts [hunk 5]
```diff
@@ -4,18 +4,18 @@ export interface NamedComponent {
 	componentName: string
 }
 
-export type ComponentConstructor<D> = NamedComponent &
-	(new (element: HTMLElement, data: D) => Component<D>)
+export type ComponentConstructor<D, E extends HTMLElement = HTMLElement> = NamedComponent &
+	(new (element: E, data: D) => Component<D, E>)
```
Comment (enzy) [inline theme/scripts/components/Component.ts]: I'm lost, some line comment about whats happening would help

File: theme/scripts/components/Component.ts (hunk 2 above)
Comment (Onset) [reply theme/scripts/components/Component.ts]: Cc. @JirkaVebr

File: theme/scripts/components/Example.ts [hunk 6]
```diff
@@ -14,7 +14,7 @@ export default class Example extends Component<ExampleData> {
 	]
 
 	init() {
-		// Initialize your component.
+		this.getChild('.example-child', HTMLElement).innerText += ` ${this.data.name}!`
```
[the remark this answers is not in the document]
Comment (Onset) [reply theme/scripts/components/Example.ts]: e.g. `HTMLInputElement`

[the remark this answers is not in the document]
Comment (enzy) [reply theme/scripts/components/Example.ts]: You can do that by modifying the selector to `input.example-child` (and even more by `[attr]` and other advanced selectors). So my question is if this part is even needed?

[the remark this answers is not in the document]
Comment (Onset) [reply theme/scripts/components/Example.ts]: I'm not sure what you mean. `input` in the selector won't change a thing. Typescript won't even recognize it as a `HTMLElement`.

image

[the remark this answers is not in the document]
Comment (enzy) [reply theme/scripts/components/Example.ts]: So the second argument is only for the type casting then?

[the remark this answers is not in the document]
Comment (Onset) [reply theme/scripts/components/Example.ts]: Correct.

Jirka is working on a default value for the second argument. https://github.com/manGoweb/MangoPress/pull/173#discussion_r231123550

[the remark this answers is not in the document]
Comment (enzy) [reply theme/scripts/components/Example.ts]: Can we add some JSDoc instructions about the behaviour please?

[the remark this answers is not in the document]
Comment (Onset) [reply theme/scripts/components/Example.ts]: I believe that's out of scope of this PR since JSDoc isn't used anywhere in MangoPress.

**Pull request merged**